### PR TITLE
RM-7635 Document how the TypeScript build infrastructure works

### PR DESCRIPTION
**Pull request by a-ctor**
https://re-motion.atlassian.net/browse/RM-7635

### CODE REVIEW

File: Remotion/Web/ClientScript/README.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+Instead, there are two additional `tsconfig.json` files located in `~/tsconfig/debug` and `~/tsconfig/release`. These inherit from the `tsconfig.json` in the root directory and contain configuration specific settings. Their output is emitted in the `bin/Debug/dist` or `bin/Release/dist` folders. This convoluted folder setup is necessary as the `Microsoft.TypeScript.MSBuild` Nuget package can searches for files named `tsconfig.json`, which makes it impossible to have different configurations for Debug/Release in the same folder. `JavaScriptLibrary.props` adds the correct `tsconfig.json` file as `Content` depending on the current build configuration.
+
+## Output files
+
+As previously mentioned, the build outputs are located in `bin/<Configuration>/dist` and there are a few different kind of output files. Only the `*.js` files are actually necessary for downstream projects but there are other kinds of file outputted as well:
```
Comment (Frazerus) [inline Remotion/Web/ClientScript/README.md:26]:
```suggestion
There are different kinds of output files generated in the build output directories `bin/<Configuration>/dist`. Only the `*.js` files are necessary for downstream projects, the other file types have other usecases:
```

File: Remotion/Web/ClientScript/README.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+- `*.js.map` files provide source mapping support for the generated JavaScript files. Serving these with the JavaScript files will allow browsers to display the TypeScript code in the debugger instead of the generated JavaScript files.
+- `*.d.ts` files provide type definitions for the generated files. Downstream projects can reference these if they themselves have a TypeScript build setup and need to reference functions/types from the generated JavaScript files.
+- `*.tsbuildinfo` files are only used for project references (see section below) and do not need to be shipped.
+
+## Project references
+
+References between different TypeScript projects are done using TypeScript's project references. The `Remotion.ObjectBinding.Web.ClientScript` project, for example, has a project reference on `Remotion.Web.ClientScript`, which can be found in all three `tsconfig.json` files. The Debug/Release files reference their respective `tsconfig.json` in the other project while the root "IDE" `tsconfig.json` reference the debug version in the other project. This is necessary as it is not possible to reference the other project's "IDE" `tsconfig.json` as it does not emit any outputs.
```
Comment (Frazerus) [inline Remotion/Web/ClientScript/README.md:34]:
```suggestion
References between different TypeScript projects are done using TypeScript's project references. The `Remotion.ObjectBinding.Web.ClientScript` project, for example, has a project reference on `Remotion.Web.ClientScript`, which can be found in all three `tsconfig.json` files. The Debug/Release files reference their respective `tsconfig.json` in the other project. The root "IDE" `tsconfig.json` references the debug version in the other project. This is necessary as it is not possible to reference the other project's "IDE" `tsconfig.json` as it does not emit any output.
```

File: Remotion/Web/ClientScript/README.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+For project reference to work there are requirements in the referenced project and the referencing project. The references project must have `"composite": true` and must have a `baseUrl` set. The referencing project on the other hand has the `"references": { "path": "..." }` entry and must also have the MSBuild property `<TypeScriptBuildMode>true</TypeScriptBuildMode>` set.
```
Comment (Frazerus) [inline Remotion/Web/ClientScript/README.md:36]:
```suggestion
There are requirements for the referenced and the referencing project for project references to work. The referenced project must have `"composite": true` and must have a `baseUrl` set. The referencing project has the `"references": { "path": "..." }` entry and must also have the MSBuild property `<TypeScriptBuildMode>true</TypeScriptBuildMode>` set.
```

**Review by Frazerus — CHANGES_REQUESTED** [review]

File: Remotion/Web/ClientScript/README.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+- `Remotion.Web.ClientScript`
+- `Remotion.ObjectBinding.Web.ClientScript` (references `Remotion.Web.ClientScript`)
+
+## General Setup
+
+To simplify the integration into the C# build, TS projects are also modelled as C# projects but their C# output is not used. Such a project have the `JavaScriptLibrary` project type.
+
+TypeScript compilation is enabled using the `Microsoft.TypeScript.MSBuild` Nuget package. While it is possible to configure TypeScript properties using MSBuild properties, we opted to use `tsconfig.json` files instead. The `Microsoft.TypeScript.MSBuild` Nuget package will build any `tsconfig.json` file that is added to MSBuild item group `Content`.
+
+The `Scripts` folder contains the TypeScript files that are compiled as part of the project. There are regular TypeScript files `*.ts` and TypeScript definition files `*.d.ts`, with the definition files being located in the `typings` subfolder. `common.d.ts` defines common types and type helpers that we created, while further subfolders contains typings for certain libraries (e.g. microsoft-ajax).
```
[the remark this answers is not in the document]
Comment (a-ctor) [reply Remotion/Web/ClientScript/README.md:16]: Do you mean the `"./Scripts/**/*.js"` include in the `tsconfig.json`? Yeah we could clean that up but I wouldn't do it in this task.

[the remark this answers is not in the document]
Comment (Frazerus) [reply Remotion/Web/ClientScript/README.md:16]: Yeah, I didn't see that we had a few other TS tasks lined up as well.

**Review by MichaelKetting — APPROVED** [review]
It looks okay to me. I noticed that I knew everything that's written in there by sight, so I cannot say if someone who isn't aware of these things can start working based on the didactic presentation. Let's assume that this is indeed the case and if martin has to fix some JS next year, we will have a live beta test of the docs :)

**Pull request merged**